MemStore locking enhancements and bufferlist alternative

cbodley opened the pull request:
Replaces MemStore's global apply_lock with new per-Sequencer locks.
Adds finer-grained object locks in order to reduce the scope of the Collection's rwlock.
Introduces PageSet, a tree-based alternative to bufferlist for storing object data.

liewegas commented:
This looks good to me!

yuyuyu101 commented on src/os/MemStore.h:
```diff
@@ -16,39 +16,58 @@
-    bufferlist data;
+  struct Object : public RefCountedObject {
+    std::mutex xattr_mutex;
+    std::mutex omap_mutex;
```
use common/Mutex instead?

cbodley replied:
Thanks for the review. If you take a look at MemStore::_clone() and it's use of std::lock() to safely take the locks of multiple objects, that's something that the Mutex interface doesn't provide.

yuyuyu101 replied:
Yep, the only thing we need to consider is Mutex provides with deadlock detect. And a wrapper for lock is good to debug or for the future purpose

tchaikov commented on src/test/CMakeLists.txt:
```diff
@@ -223,6 +223,11 @@ add_executable(kvstorebench
   ${TCMALLOC_LIBS})
 
+# osbench
+add_executable(osbench objectstore_bench.cc
```
@cbodley cool! seems we are deprecating autotools. =) but it would be ideal to at least see if `osbench` and `unittest_pageset` compile or not on supported platforms before dropping the support of the autotools.

BTW, if you add `osbench` to Makefile.am, please be sure to update `src/.gitignore` as well. otherwise the gitbuilder will be annoyed: the build script expects a "clean" repo.

cbodley commented:
push v2 makes the following changes:
-rename osbench to ceph_objectstore_bench
-added Makefile targets for ceph_objectstore_bench and unittest_pageset
-used strict_sistrtoll() to remove duplicated code
-PageSetObject::read() preallocates its buffer
-added configurable page size and config variable memstore_page_size

noahdesu commented:
Does this PR correspond to http://tracker.ceph.com/issues/12719 ? I'm getting the following traceback when starting this branch with `./vstart.sh -x -n -l --memstore`. The OSD immediately dies:

```
 ceph version 9.0.3-872-gf3b4530 (f3b45300986f9974379555888db778e9821c9386)
 1: (()+0x69b09d) [0x7f3eb7db209d]
 2: (()+0xfcb0) [0x7f3eb6a1fcb0]
 3: (gsignal()+0x35) [0x7f3eb4eec0d5]
 4: (abort()+0x17b) [0x7f3eb4eef83b]
 5: (__gnu_cxx::__verbose_terminate_handler()+0x15d) [0x7f3eb57e916d]
 6: (()+0x5f1d6) [0x7f3eb57e71d6]
 7: (()+0x5f221) [0x7f3eb57e7221]
 8: (()+0x5f438) [0x7f3eb57e7438]
 9: (()+0x794882) [0x7f3eb7eab882]
 10: (OSDSuperblock::decode(ceph::buffer::list::iterator&)+0x29) [0x7f3eb7ad5409]
 11: (OSD::read_superblock()+0x268) [0x7f3eb79a4b88]
 12: (OSD::init()+0x353) [0x7f3eb79f43d3]
 13: (main()+0x2811) [0x7f3eb7991471]
 14: (__libc_start_main()+0xed) [0x7f3eb4ed776d]
 15: (()+0x280aa9) [0x7f3eb7997aa9]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```

cbodley commented:
push v3 makes the following changes:
-fixed a bug that resulted in zero-length reads (thanks @noahdesu @loic-bot)
-fixed type conversions related to size_t for trusty-i386 build
-added ceph_objectstore_bench to ceph.spec.in to make centos7 gitbuilder stop complaining about  "Installed (but unpackaged) file(s) found"

cbodley commented:
test_objectstore now passes 28/30, but fails  SimpleCloneTest/0 and SimpleCloneRangeTest/0 due to MemStore::PageSetObject::clone() returning -ENOTSUP. All 30 tests pass if run with --memstore_page_set=0.

cbodley commented:
MemStore::PageSetObject::clone() is now implemented, and passes all test_objectstore tests

yuyuyu101 commented:
@liewegas @jdurgin @dillaman do you think it worth to use PageSet in ObjectCacher? It seemed to me that ObjectCacher also has the same problem presented in https://github.com/ceph/ceph/commit/4e6548ca750601b01c4c1d035ca81505dca94fbf

liewegas commented:
On Mon, 31 Aug 2015, Haomai Wang wrote:
> @liewegas @jdurgin @dillaman do you think it worth to use PageSet in ObjectCacher? It seemed to me that ObjectCacher also has the same problem
> presented in 4e6548c

On it's surface that sounds like a great idea to me!

jdurgin commented:
@yuyuyu101 yeah, I'd guess it would help in ObjectCacher too. Making ObjectCacher locking more fine-grained might be needed to see larger benefits from it
